### PR TITLE
[IOTDB-3081] Implementation of SlidingWindowAggregationOperator

### PR DESCRIPTION
**Pull request by liuminghui233**
In this PR, I implementation of SlidingWindowAggregationOperator.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/SmoothQueueSlidingWindowAggregator.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  protected void evictingExpiredValue() {
+    while (!deque.isEmpty() && !curTimeRange.contains(deque.getFirst().getTime())) {
+      PartialAggregationResult partialResult = deque.removeFirst();
+      if (deque.isEmpty()) {
+        this.accumulator.reset();
+      } else {
+        this.accumulator.addIntermediate(partialResult.opposite());
+      }
+    }
+  }
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/SmoothQueueSlidingWindowAggregator.java:48]:
```suggestion
    if (!deque.isEmpty() && !curTimeRange.contains(deque.getLast().getTime())) {
      this.accumulator.reset();
      return;
    }
    while (!deque.isEmpty() && !curTimeRange.contains(deque.getFirst().getTime())) {
      PartialAggregationResult partialResult = deque.removeFirst();
      this.accumulator.addIntermediate(partialResult.opposite());
    }
```

File: server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/NormalQueueSlidingWindowAggregator.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+  @Override
+  protected void evictingExpiredValue() {
+    while (!deque.isEmpty() && !curTimeRange.contains(deque.getFirst().getTime())) {
+      deque.removeFirst();
+    }
+    this.accumulator.reset();
+    if (!deque.isEmpty()) {
+      this.accumulator.addIntermediate(deque.getFirst().getPartialResult());
+    }
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/NormalQueueSlidingWindowAggregator.java:48]: Is it necessary to update the value here?

File: server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/SlidingWindowAggregator.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  /** evicting expired element in queue and reset expired aggregateResult */
+  protected abstract void evictingExpiredValue();
+
+  /** update queue and aggregateResult */
+  public abstract void processPartialResult(PartialAggregationResult partialResult);
+
+  protected static class PartialAggregationResult {
+
+    private final TimeColumn timeColumn;
+    private final Column[] valueColumns;
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/SlidingWindowAggregator.java:86]: Maybe it should be renamed to `partialResultColumns`. When I read the code, I always thought the valueColumns means the value of origin time series.

**Review by Alima777 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/SlidingWindowAggregator.java (hunk 3 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/SlidingWindowAggregator.java:86]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/NormalQueueSlidingWindowAggregator.java (hunk 2 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/NormalQueueSlidingWindowAggregator.java:48]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/SmoothQueueSlidingWindowAggregator.java (hunk 1 above)
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/SmoothQueueSlidingWindowAggregator.java:48]: fixed

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/SlidingWindowAggregationOperator.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    TsBlock inputTsBlock = cachedTsBlock;
+    while (!calcFromTsBlock(inputTsBlock, curTimeRange)) {
+      if (child.hasNext()) {
+        inputTsBlock = child.next();
+      } else {
+        break;
+      }
+    }
+    if (inputTsBlock != null && !satisfied(inputTsBlock, curTimeRange, ascending)) {
+      cachedTsBlock = inputTsBlock;
```
[the remark this answers is not in the document]
Comment (liuminghui233) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/SlidingWindowAggregationOperator.java:105]: fixed

**Pull request merged**